File: packages/domain/src/hyrox.ts
import type { TransitionMap } from './shared/machine';
import type { IsoDate } from './shared/time';

// ── Exercise library (blueprint §44) ────────────────────────────────────────
export type ExerciseCategory =
  | 'ERG'
  | 'SLED'
  | 'JUMP'
  | 'CARRY'
  | 'LUNGE'
  | 'THROW'
  | 'RUN'
  | 'CONDITIONING';

export interface Exercise {
  id: string;
  name: string;
  category: ExerciseCategory;
  equipment: string[];
  /** 1–8 when the exercise IS one of the HYROX race stations. */
  hyroxStationOrder: number | null;
  difficulty: 1 | 2 | 3;
  defaultSpec: { distanceM: number | null; reps: number | null };
  /** How-to video (YouTube) shown from the workout player. */
  videoUrl: string | null;
}

/** Substitution modeled as data (blueprint §45). */
export interface SubstitutionRule {
  originalExerciseId: string;
  alternativeExerciseId: string;
  /** 0–1: muscle + energy-system similarity. */
  similarity: number;
  conversionNote: string;
}

export function listSubstitutes(
  exerciseId: string,
  substitutions: readonly SubstitutionRule[],
  exercises: readonly Exercise[],
): { exercise: Exercise; rule: SubstitutionRule }[] {
  return substitutions
    .filter((s) => s.originalExerciseId === exerciseId)
    .sort((a, b) => b.similarity - a.similarity)
    .map((rule) => ({
      rule,
      exercise: exercises.find((e) => e.id === rule.alternativeExerciseId)!,
    }))
    .filter((x) => x.exercise !== undefined);
}

// ── Generator (blueprint §41–43, 46–47) ─────────────────────────────────────
export const WORKOUT_TYPES = ['FULL_SIMULATION', 'COVERAGE', 'QUICK', 'PRACTICE'] as const;
export type WorkoutType = (typeof WORKOUT_TYPES)[number];

export const DIVISIONS = ['MEN_OPEN', 'MEN_PRO', 'WOMEN_OPEN', 'WOMEN_PRO'] as const;
export type Division = (typeof DIVISIONS)[number];

export interface WorkoutBlock {
  order: number;
  kind: 'RUN' | 'STATION';
  exerciseId: string;
  exerciseName: string;
  /** Set when the block is a substitution for an excluded/unavailable station. */
  originalExerciseId: string | null;
  originalExerciseName: string | null;
  distanceM: number | null;
  reps: number | null;
  weightNote: string | null;
  targetSec: number;
}

export interface GeneratedWorkout {
  id: string;
  memberId: string;
  type: WorkoutType;
  division: Division;
  blocks: WorkoutBlock[];
  excludedExerciseIds: string[];
  totalTargetSec: number;
  createdAt: IsoDate;
}

/** Division-specific loads for the loaded stations (race-accurate notes). */
const WEIGHT_NOTES: Record<number, Partial<Record<Division, string>>> = {
  2: { MEN_OPEN: 'Sled 152 kg', MEN_PRO: 'Sled 202 kg', WOMEN_OPEN: 'Sled 102 kg', WOMEN_PRO: 'Sled 152 kg' },
  3: { MEN_OPEN: 'Sled 103 kg', MEN_PRO: 'Sled 153 kg', WOMEN_OPEN: 'Sled 78 kg', WOMEN_PRO: 'Sled 103 kg' },
  6: { MEN_OPEN: '2×24 kg', MEN_PRO: '2×32 kg', WOMEN_OPEN: '2×16 kg', WOMEN_PRO: '2×24 kg' },
  7: { MEN_OPEN: '20 kg bag', MEN_PRO: '30 kg bag', WOMEN_OPEN: '10 kg bag', WOMEN_PRO: '20 kg bag' },
  8: { MEN_OPEN: '6 kg ball', MEN_PRO: '9 kg ball', WOMEN_OPEN: '4 kg ball', WOMEN_PRO: '6 kg ball' },
};

/** Wall balls rep count differs per division. */
const WALL_BALL_REPS: Record<Division, number> = {
  MEN_OPEN: 100,
  MEN_PRO: 100,
  WOMEN_OPEN: 75,
  WOMEN_PRO: 100,
};

/** Rough target seconds per full station effort, per division tier. */
const STATION_TARGET_SEC: Record<number, number> = {
  1: 240, 2: 180, 3: 210, 4: 270, 5: 250, 6: 120, 7: 260, 8: 300,
};
const RUN_PACE_SEC_PER_KM: Record<Division, number> = {
  MEN_OPEN: 330,
  MEN_PRO: 300,
  WOMEN_OPEN: 360,
  WOMEN_PRO: 330,
};
const PRO_FACTOR: Record<Division, number> = {
  MEN_OPEN: 1,
  MEN_PRO: 0.9,
  WOMEN_OPEN: 1.05,
  WOMEN_PRO: 0.95,
};

export interface GenerateWorkoutArgs {
  type: WorkoutType;
  division: Division;
  /** COVERAGE/PRACTICE: which stations (1–8) to include; empty → picked via `pick`. */
  stationOrders: number[];
  excludedExerciseIds: string[];
  exercises: readonly Exercise[];
  substitutions: readonly SubstitutionRule[];
  /** Deterministic chooser: returns an int in [0, n). */
  pick: (n: number) => number;
}

export function generateWorkout(
  args: GenerateWorkoutArgs,
): { blocks: WorkoutBlock[]; totalTargetSec: number } {
  const { division, exercises, substitutions } = args;
  const stations = exercises
    .filter((e) => e.hyroxStationOrder !== null)
    .sort((a, b) => a.hyroxStationOrder! - b.hyroxStationOrder!);
  const running = exercises.find((e) => e.category === 'RUN');
  const blocks: WorkoutBlock[] = [];
  let order = 1;

  const resolveStation = (station: Exercise): { exercise: Exercise; original: Exercise | null } => {
    if (!args.excludedExerciseIds.includes(station.id)) return { exercise: station, original: null };
    const sub = listSubstitutes(station.id, substitutions, exercises).find(
      (s) => !args.excludedExerciseIds.includes(s.exercise.id),
    );
    return sub ? { exercise: sub.exercise, original: station } : { exercise: station, original: null };
  };

  const pushRun = (distanceM: number) => {
    blocks.push({
      order: order++,
      kind: 'RUN',
      exerciseId: running?.id ?? 'ex_run',
      exerciseName: 'Running',
      originalExerciseId: null,
      originalExerciseName: null,
      distanceM,
      reps: null,
      weightNote: null,
      targetSec: Math.round((RUN_PACE_SEC_PER_KM[division] * distanceM) / 1000),
    });
  };

  const pushStation = (station: Exercise, volumeFactor: number) => {
    const { exercise, original } = resolveStation(station);
    const stationOrder = station.hyroxStationOrder!;
    const isWallBalls = stationOrder === 8;
    const baseReps = isWallBalls ? WALL_BALL_REPS[division] : station.defaultSpec.reps;
    blocks.push({
      order: order++,
      kind: 'STATION',
      exerciseId: exercise.id,
      exerciseName: exercise.name,
      originalExerciseId: original?.id ?? null,
      originalExerciseName: original?.name ?? null,
      distanceM: station.defaultSpec.distanceM
        ? Math.round(station.defaultSpec.distanceM * volumeFactor)
        : null,
      reps: baseReps ? Math.round(baseReps * volumeFactor) : null,
      weightNote: WEIGHT_NOTES[stationOrder]?.[division] ?? null,
      targetSec: Math.round(STATION_TARGET_SEC[stationOrder]! * PRO_FACTOR[division] * volumeFactor),
    });
  };

  const chooseStations = (count: number): Exercise[] => {
    if (args.stationOrders.length > 0) {
      return stations.filter((s) => args.stationOrders.includes(s.hyroxStationOrder!));
    }
    const pool = [...stations];
    const chosen: Exercise[] = [];
    while (chosen.length < count && pool.length > 0) {
      chosen.push(pool.splice(args.pick(pool.length), 1)[0]!);
    }
    return chosen.sort((a, b) => a.hyroxStationOrder! - b.hyroxStationOrder!);
  };

  switch (args.type) {
    case 'FULL_SIMULATION':
      for (const station of stations) {
        pushRun(1000);
        pushStation(station, 1);
      }
      break;
    case 'COVERAGE':
      for (const station of chooseStations(4)) {
        pushRun(600);
        pushStation(station, 1);
      }
      break;
    case 'QUICK':
      for (const station of chooseStations(4)) {
        pushRun(400);
        pushStation(station, 0.5);
      }
      break;
    case 'PRACTICE': {
      const target = chooseStations(1)[0] ?? stations[0]!;
      for (let round = 0; round < 3; round++) {
        pushRun(200);
        pushStation(target, 0.5);
      }
      break;
    }
  }

  return { blocks, totalTargetSec: blocks.reduce((s, b) => s + b.targetSec, 0) };
}

/** Customize step (§46): swap one station block for a substitute exercise. */
export function replaceBlockExercise(
  block: WorkoutBlock,
  replacement: Exercise,
): WorkoutBlock {
  return {
    ...block,
    exerciseId: replacement.id,
    exerciseName: replacement.name,
    originalExerciseId: block.originalExerciseId ?? block.exerciseId,
    originalExerciseName: block.originalExerciseName ?? block.exerciseName,
  };
}

// ── Active workout session (blueprint §48–49) ───────────────────────────────
export const WORKOUT_SESSION_STATUSES = [
  'READY',
  'STARTED',
  'PAUSED',
  'COMPLETED',
  'PARTIAL',
] as const;
export type WorkoutSessionStatus = (typeof WORKOUT_SESSION_STATUSES)[number];

export const WORKOUT_SESSION_TRANSITIONS: TransitionMap<WorkoutSessionStatus> = {
  READY: ['STARTED'],
  STARTED: ['PAUSED', 'COMPLETED', 'PARTIAL'],
  PAUSED: ['STARTED', 'PARTIAL'],
  COMPLETED: [],
  PARTIAL: [],
};

export interface WorkoutBlockResult {
  order: number;
  durationSec: number;
}

export interface WorkoutSession {
  id: string;
  workoutId: string;
  memberId: string;
  status: WorkoutSessionStatus;
  /** Order of the block currently being executed. */
  currentBlock: number;
  startedAt: IsoDate | null;
  endedAt: IsoDate | null;
  blockResults: WorkoutBlockResult[];
  pauseCount: number;
  totalPauseSec: number;
  createdAt: IsoDate;
}

export function sessionActiveSec(session: Pick<WorkoutSession, 'blockResults'>): number {
  return session.blockResults.reduce((s, r) => s + r.durationSec, 0);
}

export function sessionCompletionPct(
  session: Pick<WorkoutSession, 'blockResults'>,
  totalBlocks: number,
): number {
  if (totalBlocks === 0) return 0;
  return Math.round((session.blockResults.length / totalBlocks) * 100);
}

File: apps/backend/internal/domain/hyrox.go
package domain

import (
	"math"
	"sort"
	"time"
)

// The HYROX side of training: generating a workout from the exercise library,
// and running one.
//
// A race is eight one-kilometre runs, each followed by a station, always in the
// same order. Everything here is derived from that one fact — the shorter
// formats are the same shape with fewer stations and less of each.

// WorkoutType is how much of a race the session covers.
type WorkoutType string

const (
	WorkoutFullSimulation WorkoutType = "FULL_SIMULATION"
	WorkoutCoverage       WorkoutType = "COVERAGE"
	WorkoutQuick          WorkoutType = "QUICK"
	WorkoutPractice       WorkoutType = "PRACTICE"
)

func IsValidWorkoutType(value string) bool {
	switch WorkoutType(value) {
	case WorkoutFullSimulation, WorkoutCoverage, WorkoutQuick, WorkoutPractice:
		return true
	}
	return false
}

// Division sets the loads and the rep counts.
type Division string

const (
	DivisionMenOpen   Division = "MEN_OPEN"
	DivisionMenPro    Division = "MEN_PRO"
	DivisionWomenOpen Division = "WOMEN_OPEN"
	DivisionWomenPro  Division = "WOMEN_PRO"
)

func IsValidDivision(value string) bool {
	switch Division(value) {
	case DivisionMenOpen, DivisionMenPro, DivisionWomenOpen, DivisionWomenPro:
		return true
	}
	return false
}

// WorkoutBlock is one run or one station.
type WorkoutBlock struct {
	Order        int    `json:"order"`
	Kind         string `json:"kind"`
	ExerciseID   string `json:"exerciseId"`
	ExerciseName string `json:"exerciseName"`
	// Original* is set when this block stands in for a station the member
	// excluded, so the player can say what it is replacing.
	OriginalExerciseID   *string `json:"originalExerciseId"`
	OriginalExerciseName *string `json:"originalExerciseName"`
	DistanceM            *int    `json:"distanceM"`
	Reps                 *int    `json:"reps"`
	WeightNote           *string `json:"weightNote"`
	TargetSec            int     `json:"targetSec"`
}

// GeneratedWorkout is a workout as authored: read as a unit, never queried
// into, which is why the blocks live in one JSON column.
type GeneratedWorkout struct {
	ID                  string         `json:"id"`
	MemberID            string         `json:"memberId"`
	Type                WorkoutType    `json:"type"`
	Division            Division       `json:"division"`
	Blocks              []WorkoutBlock `json:"blocks"`
	ExcludedExerciseIDs []string       `json:"excludedExerciseIds"`
	TotalTargetSec      int            `json:"totalTargetSec"`
	CreatedAt           time.Time      `json:"createdAt"`
}

// Race-accurate loads for the stations that carry one.
var stationWeightNotes = map[int]map[Division]string{
	2: {DivisionMenOpen: "Sled 152 kg", DivisionMenPro: "Sled 202 kg", DivisionWomenOpen: "Sled 102 kg", DivisionWomenPro: "Sled 152 kg"},
	3: {DivisionMenOpen: "Sled 103 kg", DivisionMenPro: "Sled 153 kg", DivisionWomenOpen: "Sled 78 kg", DivisionWomenPro: "Sled 103 kg"},
	6: {DivisionMenOpen: "2×24 kg", DivisionMenPro: "2×32 kg", DivisionWomenOpen: "2×16 kg", DivisionWomenPro: "2×24 kg"},
	7: {DivisionMenOpen: "20 kg bag", DivisionMenPro: "30 kg bag", DivisionWomenOpen: "10 kg bag", DivisionWomenPro: "20 kg bag"},
	8: {DivisionMenOpen: "6 kg ball", DivisionMenPro: "9 kg ball", DivisionWomenOpen: "4 kg ball", DivisionWomenPro: "6 kg ball"},
}

// Wall balls is the one station whose rep count changes with the division.
var wallBallReps = map[Division]int{
	DivisionMenOpen: 100, DivisionMenPro: 100, DivisionWomenOpen: 75, DivisionWomenPro: 100,
}

// Target seconds for a full effort at each station, before the division factor.
var stationTargetSec = map[int]int{1: 240, 2: 180, 3: 210, 4: 270, 5: 250, 6: 120, 7: 260, 8: 300}

var runPaceSecPerKm = map[Division]int{
	DivisionMenOpen: 330, DivisionMenPro: 300, DivisionWomenOpen: 360, DivisionWomenPro: 330,
}

// proFactor scales the station targets: a pro field is faster over heavier
// loads, and an open women's field is a little slower over lighter ones.
var proFactor = map[Division]float64{
	DivisionMenOpen: 1, DivisionMenPro: 0.9, DivisionWomenOpen: 1.05, DivisionWomenPro: 0.95,
}

// GenerateWorkoutArgs is everything the generator needs.
type GenerateWorkoutArgs struct {
	Type     WorkoutType
	Division Division
	// StationOrders picks the stations for the shorter formats. Empty means
	// "choose for me", and Pick does the choosing.
	StationOrders       []int
	ExcludedExerciseIDs []string
	Exercises           []Exercise
	Substitutions       []SubstitutionRule
	// Pick returns an int in [0, n). It is injected so a test gets the same
	// workout every run and the service can hand it a real random source.
	Pick func(n int) int
}

// ListSubstitutes is what can stand in for an exercise, best match first.
func ListSubstitutes(exerciseID string, substitutions []SubstitutionRule, exercises []Exercise) []Exercise {
	byID := make(map[string]Exercise, len(exercises))
	for _, e := range exercises {
		byID[e.ID] = e
	}
	rules := []SubstitutionRule{}
	for _, r := range substitutions {
		if r.OriginalExerciseID == exerciseID {
			if _, ok := byID[r.AlternativeExerciseID]; ok {
				rules = append(rules, r)
			}
		}
	}
	sort.SliceStable(rules, func(i, j int) bool { return rules[i].Similarity > rules[j].Similarity })
	out := make([]Exercise, 0, len(rules))
	for _, r := range rules {
		out = append(out, byID[r.AlternativeExerciseID])
	}
	return out
}

// GenerateWorkout builds the blocks for a session.
//
// Excluded stations are substituted rather than dropped: somebody without a
// sled still needs the work that station does, and a session with a hole in it
// teaches the wrong pacing.
func GenerateWorkout(args GenerateWorkoutArgs) ([]WorkoutBlock, int) {
	stations := []Exercise{}
	var running *Exercise
	for i, e := range args.Exercises {
		if e.HyroxStationOrder != nil {
			stations = append(stations, e)
		}
		if e.Category == CategoryRun && running == nil {
			running = &args.Exercises[i]
		}
	}
	sort.SliceStable(stations, func(i, j int) bool {
		return *stations[i].HyroxStationOrder < *stations[j].HyroxStationOrder
	})

	excluded := make(map[string]bool, len(args.ExcludedExerciseIDs))
	for _, id := range args.ExcludedExerciseIDs {
		excluded[id] = true
	}

	blocks := []WorkoutBlock{}
	order := 1

	pushRun := func(distanceM int) {
		id, name := "ex_run", "Running"
		if running != nil {
			id = running.ID
		}
		blocks = append(blocks, WorkoutBlock{
			Order: order, Kind: "RUN", ExerciseID: id, ExerciseName: name,
			DistanceM: intPtr(distanceM),
			TargetSec: int(math.Round(float64(runPaceSecPerKm[args.Division]*distanceM) / 1000)),
		})
		order++
	}

	pushStation := func(station Exercise, volume float64) {
		exercise, original := station, (*Exercise)(nil)
		if excluded[station.ID] {
			for _, sub := range ListSubstitutes(station.ID, args.Substitutions, args.Exercises) {
				if !excluded[sub.ID] {
					exercise, original = sub, &station
					break
				}
			}
		}
		stationOrder := *station.HyroxStationOrder

		block := WorkoutBlock{
			Order: order, Kind: "STATION",
			ExerciseID: exercise.ID, ExerciseName: exercise.Name,
			TargetSec: int(math.Round(float64(stationTargetSec[stationOrder]) * proFactor[args.Division] * volume)),
		}
		if original != nil {
			block.OriginalExerciseID = &original.ID
			block.OriginalExerciseName = &original.Name
		}
		if station.DefaultSpec.DistanceM != nil {
			block.DistanceM = intPtr(int(math.Round(float64(*station.DefaultSpec.DistanceM) * volume)))
		}
		reps := station.DefaultSpec.Reps
		if stationOrder == 8 {
			reps = intPtr(wallBallReps[args.Division])
		}
		if reps != nil {
			block.Reps = intPtr(int(math.Round(float64(*reps) * volume)))
		}
		if note, ok := stationWeightNotes[stationOrder][args.Division]; ok {
			block.WeightNote = &note
		}
		blocks = append(blocks, block)
		order++
	}

	chooseStations := func(count int) []Exercise {
		if len(args.StationOrders) > 0 {
			wanted := make(map[int]bool, len(args.StationOrders))
			for _, o := range args.StationOrders {
				wanted[o] = true
			}
			chosen := []Exercise{}
			for _, s := range stations {
				if wanted[*s.HyroxStationOrder] {
					chosen = append(chosen, s)
				}
			}
			return chosen
		}
		pool := append([]Exercise{}, stations...)
		chosen := []Exercise{}
		for len(chosen) < count && len(pool) > 0 {
			i := args.Pick(len(pool))
			chosen = append(chosen, pool[i])
			pool = append(pool[:i], pool[i+1:]...)
		}
		sort.SliceStable(chosen, func(i, j int) bool {
			return *chosen[i].HyroxStationOrder < *chosen[j].HyroxStationOrder
		})
		return chosen
	}

	switch args.Type {
	case WorkoutFullSimulation:
		// The race itself: every station, full volume, a kilometre between each.
		for _, station := range stations {
			pushRun(1000)
			pushStation(station, 1)
		}
	case WorkoutCoverage:
		for _, station := range chooseStations(4) {
			pushRun(600)
			pushStation(station, 1)
		}
	case WorkoutQuick:
		for _, station := range chooseStations(4) {
			pushRun(400)
			pushStation(station, 0.5)
		}
	case WorkoutPractice:
		// One station, three times: this is the format for fixing something.
		target := stations
		if chosen := chooseStations(1); len(chosen) > 0 {
			target = chosen[:1]
		}
		if len(target) > 0 {
			for range 3 {
				pushRun(200)
				pushStation(target[0], 0.5)
			}
		}
	}

	total := 0
	for _, b := range blocks {
		total += b.TargetSec
	}
	return blocks, total
}

// ReplaceBlockExercise swaps one block's exercise, remembering what it was.
//
// The original is only recorded the first time: swapping twice still says what
// the block is standing in for at the race, not what it was a minute ago.
func ReplaceBlockExercise(block WorkoutBlock, replacement Exercise) WorkoutBlock {
	if block.OriginalExerciseID == nil {
		id, name := block.ExerciseID, block.ExerciseName
		block.OriginalExerciseID = &id
		block.OriginalExerciseName = &name
	}
	block.ExerciseID = replacement.ID
	block.ExerciseName = replacement.Name
	return block
}

// ── Running a session ────────────────────────────────────────────────────────

// WorkoutSessionStatus is where a session has got to.
type WorkoutSessionStatus string

const (
	WorkoutReady     WorkoutSessionStatus = "READY"
	WorkoutStarted   WorkoutSessionStatus = "STARTED"
	WorkoutPaused    WorkoutSessionStatus = "PAUSED"
	WorkoutCompleted WorkoutSessionStatus = "COMPLETED"
	WorkoutPartial   WorkoutSessionStatus = "PARTIAL"
)

// WorkoutSessionTransitions: a finished session is finished. Somebody who
// stopped halfway gets PARTIAL, which keeps the blocks they did without
// claiming they did the workout.
var WorkoutSessionTransitions = TransitionMap[WorkoutSessionStatus]{
	WorkoutReady:     {WorkoutStarted},
	WorkoutStarted:   {WorkoutPaused, WorkoutCompleted, WorkoutPartial},
	WorkoutPaused:    {WorkoutStarted, WorkoutPartial},
	WorkoutCompleted: {},
	WorkoutPartial:   {},
}

// WorkoutBlockResult is how long one block actually took.
type WorkoutBlockResult struct {
	Order       int `json:"order"`
	DurationSec int `json:"durationSec"`
}

// WorkoutSession is one attempt at a workout.
type WorkoutSession struct {
	ID            string               `json:"id"`
	WorkoutID     string               `json:"workoutId"`
	MemberID      string               `json:"memberId"`
	Status        WorkoutSessionStatus `json:"status"`
	CurrentBlock  int                  `json:"currentBlock"`
	StartedAt     *time.Time           `json:"startedAt"`
	EndedAt       *time.Time           `json:"endedAt"`
	BlockResults  []WorkoutBlockResult `json:"blockResults"`
	PauseCount    int                  `json:"pauseCount"`
	TotalPauseSec int                  `json:"totalPauseSec"`
	CreatedAt     time.Time            `json:"createdAt"`
}

// SessionActiveSec is time on the clock, which is the sum of the blocks —
// deliberately not wall time, because a session paused for lunch did not take
// four hours.
func SessionActiveSec(results []WorkoutBlockResult) int {
	total := 0
	for _, r := range results {
		total += r.DurationSec
	}
	return total
}

// SessionCompletionPct is how much of the workout was actually done.
func SessionCompletionPct(results []WorkoutBlockResult, totalBlocks int) int {
	if totalBlocks == 0 {
		return 0
	}
	return int(math.Round(float64(len(results)) / float64(totalBlocks) * 100))
}

func intPtr(v int) *int { return &v }
